WalletWasabi.Tests/UnitTests/Blockchain/TransactionBuilding/BranchAndBoundTests.cs

```diff
@@ -96,20 +113,63 @@ public void ExactMatch_ReturnNullIfNoExactMatchFoundTest()
 		Assert.Null(selectedCoins);
 	}
 
-	/// <summary>Tests that best found solution is found when exact solution does not exist.</summary>
+	/// <summary>Tests that a best found selection is found when an exact solution does not exist.</summary>
+	[Fact]
+	public void CheapestSelection_NoInputCosts()
+	{
+		List<long> inputValues = new() { 11, 7, 5, 3, 2 }; // Sum is 28.
+		long[] inputCosts = new long[] { 0, 0, 0, 0, 0 }; // No input costs. Idealized.
+		long target = 27; // Target that we cannot get as a sum of input values.
+
+		BranchAndBound algorithm = new(inputValues);
+		CheapestSelectionStrategy strategy = new(target, inputCosts);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
+
+		Assert.False(wasSuccessful);
+		Assert.Null(selectedCoins);
+		Assert.Equal(new long[] { 11, 7, 5, 3, 2 }, strategy.GetBestSelectionFound());
+	}
+
 	[Fact]
-	public void BestSum_NoExactSolution()
+	public void CheapestSelection_ExactMatchIsAlsoCheapest()
 	{
-		List<long> inputValues = new() { 2, 3, 5, 7, 11 }; // Sum is 28.
+		List<long> inputValues = new() { 35, 17, 10, 5, 3, 2 };
+
+		// Make the second input very expensive to spend so that it is not selected (not likely in reality).
+		long[] inputCosts = new long[] { 1, 5, 1, 1, 1, 1 };
+
+		long target = 27;
+
+		BranchAndBound algorithm = new(inputValues);
+		CheapestSelectionStrategy strategy = new(target, inputCosts);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
+
+		Assert.False(wasSuccessful);
+		Assert.Null(selectedCoins);
+
+		Assert.Equal(new long[] { 17, 10 }, strategy.GetBestSelectionFound());
+	}
+
+	[Fact]
+	public void CheapestSelection_ExactMatchIsExpensive()
+	{
+		List<long> inputValues = new() { 35, 17, 10, 5, 3, 2 };
+
+		// Make the second input very expensive to spend so that it is not selected (not likely in reality).
+		long[] inputCosts = new long[] { 1, 1000, 1, 1, 1, 1 };
+
 		long target = 27; // Target that we cannot get as a sum of input values.
 
 		BranchAndBound algorithm = new(inputValues);
-		BestSumStrategy strategy = new(target);
+		CheapestSelectionStrategy strategy = new(target, inputCosts);
 		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
 
 		Assert.False(wasSuccessful);
 		Assert.Null(selectedCoins);
-		Assert.Equal(new long[] { 11, 7, 5, 3, 2 }, strategy.GetBestSumFound());
+
+		// Selection (17, 10) is actually more expensive: 17 + 10 + 1000 + 1 = 1018.
+		// Whereas (35) costs us 35 + 1 = 36.
+		Assert.Equal(new long[] { 35 }, strategy.GetBestSelectionFound());
```

Comment:
What's the difference between this test and the one above? Apart from the second element of the `inputCosts`?

Reply:
That's basically it. I want to test that `inputCosts` are taken into account correctly.

Reply:
Maybe we can make it shorter using `[Theory]` attr. If you have an idea how to make it shorter, please suggest it here.

Reply:
Well, AFAIK we could use something like this:

```
[Theory]
[InlineData(new long[] { 1, 5, 1, 1, 1, 1 })]
[InlineData(new long[] { 1, 1000, 1, 1, 1, 1 })]
public void CheapestSelection_WithInputCosts(long[] inputCosts)
{
	List<long> inputValues = new() { 35, 17, 10, 5, 3, 2 };
	long target = 27;

	BranchAndBound algorithm = new(inputValues);
	CheapestSelectionStrategy strategy = new(target, inputCosts);
	bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);

	Assert.False(wasSuccessful);
	Assert.Null(selectedCoins);

	Assert.Equal(new long[] { 17, 10 }, strategy.GetBestSelectionFound());
}
```
Only problem with this is the last `Assert`, but you already changed that.

Reply:
So I have rewritten the code a bit and I'm not sure if we really need to save lines of code. If you think we should do it, please suggest another code suggestion. On one hand, I like shorter code better but on the other hand, I think it's good to have that code explanation, so I'm ~0 on this.

Reply:
I'm fine with the 2 separate tests, it was just a nitpick. Multiple possible "solution" in the tests worries me more.